### PR TITLE
[wasm] Add support for `$(PublishAot)`

### PR DESCRIPTION
**Pull request by radical**
Allow building for AOT to be controlled by `$(PublishAot)` also, besides
the existing `$(RunAOTCompilation)`.

### CODE REVIEW

**Comment by steveisok** [comment]
We should add this across the boards

**Comment by MichalStrehovsky** [comment]
Do you intend to replace RunAOTCompilation with PublishAot in docs? Or why is this needed?

PublishAot has extra behaviors in the SDK such as disabling generation of runtimeconfig.json, enabling an AOT Roslyn analyzer that flags code that is not AOT friendly, enabling another analyzer for single file friendly APIs, etc. It enforces restrictions that WASM might not be able to take advantage of (e.g. it flags APIs that would require gsharedvt fallbacks - in the absence of such APIs gsharedvt doesn't need to be generated, etc.).

If you still want to go with it, I think we would need SDK tests to ensure it actually works because the SDK assumes PublishAot is the CoreCLR-based AOT, same way PublishSingleFile assumes the CoreCLR-based one.

**Comment by MichalStrehovsky** [comment]
Looks like I didn't even have to comment - the SDK does not expect PublishAot for WASM:

```
        [tldfyvjy.yrg] /datadisks/disk1/work/B3FB09F2/w/A43D08E5/e/dotnet-workload/sdk/7.0.100-rc.1.22416.2/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(90,5): error NETSDK1183: Unable to optimize assemblies for Ahead of time compilation: a valid runtime package was not found. Either set the PublishAot property to false, or use a supported runtime identifier when publishing. When targeting .NET 7 or higher, make sure to restore packages with the PublishAot property set to true. [/datadisks/disk1/work/B3FB09F2/w/A43D08E5/e/tldfyvjy.yrg/publish_aot_property_Release.csproj]
```

**Comment by radical** [comment]
> Do you intend to replace RunAOTCompilation with PublishAot in docs? Or why is this needed?
> 
> PublishAot has extra behaviors in the SDK such as disabling generation of runtimeconfig.json, enabling an AOT Roslyn analyzer that flags code that is not AOT friendly, enabling another analyzer for single file friendly APIs, etc. It enforces restrictions that WASM might not be able to take advantage of (e.g. it flags APIs that would require gsharedvt fallbacks - in the absence of such APIs gsharedvt doesn't need to be generated, etc.).
> 
> If you still want to go with it, I think we would need SDK tests to ensure it actually works because the SDK assumes PublishAot is the CoreCLR-based AOT, same way PublishSingleFile assumes the CoreCLR-based one.

My understanding based on https://github.com/dotnet/sdk/issues/25392 was that `PublishAot` would be common for everyone. IIUC, there hasn't been any suggestion about `PublishAot` being only for CoreCLR-based AOT.

But if that's the case, then I can change it here.

/cc @lewing @steveisok

**Comment by MichalStrehovsky** [comment]
I see - I wasn't part of the conversation. This would need to start on the SDK side, as you can probably see from the failing test.

We'll need to internally (within the SDK) define a more nuanced property that PublishAot will map to depending on what RID is being targeted because PublishAot with UseMonoRuntime=true means a different thing from PublishAot without the other property.

**Comment by radical** [comment]
What needs to be done next?

**Comment by MichalStrehovsky** [comment]
> What needs to be done next?

If that's a question for me, I would start by looking for instances of PublishAot in the SDK repo and decide which ones should apply when PublishAot is used with Mono. I think this can only be tested here once this repo consumes such build of the SDK. It likely won't be sooner than Preview 1 of .NET 8 (this doesn't look like a change that would meet the .NET 7 bar).

File: src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets.in [hunk 1]
```diff
@@ -5,11 +5,13 @@
                                             '$(RuntimeIdentifier)' == 'browser-wasm' and
                                             '$(TargetFrameworkIdentifier)' == '.NETCoreApp' and
                                             !$([MSBuild]::VersionGreaterThanOrEquals('$(TargetFrameworkVersion)', '6.0'))">true</BrowserWorkloadDisabled>
+        <_AOTEnabled Condition="'$(RunAOTCompilation)' != ''">$(RunAOTCompilation)</_AOTEnabled>
+        <_AOTEnabled Condition="'$(_AOTEnabled)' == '' and '$(PublishAot)' != ''">$(PublishAot)</_AOTEnabled>
```
Comment (jonathanpeppers) [inline src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets.in:9]: If you have side-by-side `WorkloadManifest.targets` from .NET 6 and .NET 7, do we know if this file get imported twice?

Would a project import both:
```
C:\Program Files\dotnet\sdk-manifests\6.0.400\microsoft.net.workload.mono.toolchain\WorkloadManifest.targets
C:\Program Files\dotnet\sdk-manifests\7.0.100\microsoft.net.workload.mono.toolchain\WorkloadManifest.targets
```

Curious if we will need to deal with this in MAUI.

**Pull request closed without merging**